### PR TITLE
Type guard intersection types

### LINKED ISSUE #8911: Inferring Intersection Types
**TypeScript Version:** 

1.8.10

**Code**

```ts
// two interfaces
interface A {
  a: string;
}

interface B {
  b: string;
}

// a type guard for B
function isB(toTest: any): toTest is B {
  return toTest && toTest.b;
}

// a function that turns an A into an A & B
function union(a: A): A & B {
  if (isB(a)) {
    return a;
  } else {
    return null;
  }
}
```

**Expected behavior:**

I would expect this to compile successfully.

**Actual behavior:** 

```
test.ts(15,12): error TS2322: Type 'A' is not assignable to type 'A & B'.
  Type 'A' is not assignable to type 'B'.
    Property 'b' is missing in type 'A'.
```

Inside the if-block, tsc has "forgotten" about the A-ness of `a`. This is a useful pattern if `b` is an expensive property to compute and you want to do different things if `b` is or is not present.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #8911.

### CODE REVIEW

File: tests/cases/conformance/expressions/typeGuards/typeGuardIntersectionTypes.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+// Function to identify a given beast by detecting its features
+function identifyBeast(beast: Beast) {
+
+    // All beasts with legs
+    if (hasLegs(beast)) {
+
+        // All winged beasts with legs
+        if (hasWings(beast)) {
+            if (beast.legs === 4) { // ERROR TS2339: Property 'legs' does not exist on type 'Winged'.
```
Comment (mhegazy) [inline tests/cases/conformance/expressions/typeGuards/typeGuardIntersectionTypes.ts]: comments are misleading now.

**Comment by mhegazy** [comment]
:+1:

File: tests/baselines/reference/typeGuardIntersectionTypes.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+// a function that turns an A into an A & B
+function union(a: A): A & B | null {
+  if (isB(a)) {
+    return a;
+  } else {
+    return null;
+  }
+}
+
+// Repro from #9011
```
Comment (yortus) [inline tests/baselines/reference/typeGuardIntersectionTypes.js]: should be #9016

File: tests/baselines/reference/typeGuardIntersectionTypes.symbols [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+    return a;
+>a : Symbol(a, Decl(typeGuardIntersectionTypes.ts, 43, 15))
+
+  } else {
+    return null;
+  }
+}
+
+// Repro from #9011
```
Comment (yortus) [inline tests/baselines/reference/typeGuardIntersectionTypes.symbols]: should be #9016

**Comment by yortus** [comment]
This is a really great enhancement to type guards.

**Comment by yortus** [comment]
IMO this is a major improvement to control flow based type inference that should probably be mentioned in the roadmap and in the upcoming 'what's new in TypeScript' for 2.0.

**Pull request merged**